### ISSUE TITLE
`Run Keyword and Expect Error` cannot catch error caused by variable assignement

### ISSUE THREAD

**Issue by clarmso** [robotframework/robotframework#2363]
'Run Keyword and Expect Error' catches the errors from the commands used by the keywords, but it does not catch the the failed multi variable assignment from 'String.Split String'.  The test fails right away despite the keyword is wrapped by 'Run Keyword and Expect Error'.  Please see the attached file for an example.

[Dummy.txt](https://github.com/robotframework/robotframework/files/249025/Dummy.txt)

If this behaviour is not a bug, how can we document such a behaviour in the keyword library.  FYI: This particular keyword has the following limitations as disclosed by the documentation:

> Errors caused by invalid syntax, timeouts, or fatal exceptions are not caught by this keyword. Since Robot Framework 2.9, variable errors are caught by this keyword.

Here is the test result:

![2016-05-04_11-21-40](https://cloud.githubusercontent.com/assets/1740517/15019078/70ac1f6c-11ea-11e6-89b1-bf0f43d2c936.png)

Operating System: Windows 8 (64-bit)
Python interpreter: Python 2.7.9 on win32
Robot Framework: 3.0

**Comment by pekkaklarck** [comment]
The case that fails is considered a syntax error. You try to assing one return value to two variables. It could be argued that this shouldn't be a syntax error, but rather similar variable related error that can be caught already now. It would be somewhat cleaner if syntax errors could always be found by inspecting syntax, but in this case then error depends on the actual return value.

What do others think, should errors related to assignement that only occur at execution time be changed to VariableErrors? Totally invalid syntax like `@{a}  @{b} =  Kw` should still cause uncatchable DataErrors.

**Comment by clarmso** [comment]
@pekkaklarck Should we poll others if such an error that can be caught by "Run Keyword And Ignore Error"?  Having one or more variables in my case unassigned could generate undesirable results that can be hard to track down.

**Comment by Tattoo** [comment]
I personally feel this is syntax error, thus not catchable. If you are expecting two values and a keyword returns (surprisingly?) one, System Under Test is in my experience usually doing something it's not suppose to. Masking this kind of behaviour is usually what you do not want.

Even though it's more logic in the test cases, the issue can already be remedied, in my opinion, with capturing the output in a list variable:

```
*** Keywords ***
My keyword
    @{words}=    Split String    Hello    ${EMPTY}
    ${length}=    Get Length    ${words}
    Run keyword if    ${length} == 1    Fail    This error is catchable with expect error
    [Return]    ${words[0]}    ${words[1]}
```

**Comment by clarmso** [comment]
@Tattoo I agree on using a workaround like that one you demonstrated.  Some checks makes the implementation of the keyword more verbose, but the effort should be worth it.

@pekkaklarck Since it is a bad idea to catch this sort of runtime error by the Robot Framework, please feel free to close this issue if you like to.

**Comment by pekkaklarck** [comment]
Personally I'm pretty much +0 about changing this. I don't think

    ${foo}    ${bar} =    Keyword

is a syntax error regardless what `Keyword` returns, but same argument could be also used if `Keyword` wouldn't exist at all. Anyway, I'd be fine changing the resulting error if return values do not match to be catchable and could even consider the same with non-existing keywords. I don't care strongly either way, so if the original reporter thinks this isn't a good idea, and nobody else needs it, it is obviously easiest if we make no changes.

**Comment by pekkaklarck** [comment]
Decided to fix this inconsistency. The main reason is that #2404 changed semantics in teardowns so that they are stopped on syntax errors and I wouldn't like them to stop for problems like in this issue.

**Comment by pekkaklarck** [comment]
Implemented by b719108b54237541b02c1117b0966afd7bbcd438. Forgot to close this issue using the commit message.
